File: src/search_stock.py
# search_stock.py
#
# Copyright 2025 Flávio de Vasconcellos Corrêa
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <https://www.gnu.org/licenses/>.
#
# SPDX-License-Identifier: GPL-3.0-or-later

import gi

gi.require_version('Gtk', '4.0')
gi.require_version('Adw', '1')

from gi.repository import Gtk, GObject


@Gtk.Template(resource_path='/com/github/sheepfarm/merkato/search_stock.ui')
class MerkatoSearchStock(Gtk.Box):
    """
    Widget de busca de stocks.
    Contém um campo de entrada e botão de busca.
    """
    __gtype_name__ = 'MerkatoSearchStock'

    __gsignals__ = {
        'activate': (GObject.SignalFlags.RUN_FIRST, None, (str,)),
        'changed': (GObject.SignalFlags.RUN_FIRST, None, (str,)),
    }

    _entry: Gtk.Entry = Gtk.Template.Child()
    _button: Gtk.Button = Gtk.Template.Child()


    def __init__(self, **kwargs):
        super().__init__(**kwargs)
        self._setup_signals()
        self._update_button_state()


    def _setup_signals(self):
        """Configura os sinais do widget."""
        self._entry.connect('activate', self._on_activate)
        self._entry.connect('changed', self._on_changed)
        self._entry.connect('changed', self._on_text_changed)
        self._button.connect('clicked', self._on_activate)

    # ============== Callbacks ==============

    def _on_activate(self, widget):
        """Callback quando o usuário ativa a busca."""
        text = self._entry.get_text()
        if text.strip():
            self.emit('activate', text)


    def _on_changed(self, widget):
        """Callback quando o texto muda."""
        self.emit('changed', self._entry.get_text())


    def _on_text_changed(self, widget):
        """Callback para atualizar estado do botão."""
        self._update_button_state()

    # ============== Métodos públicos ==============

    def get_text(self) -> str:
        """
        Retorna o texto atual do campo de entrada.

        Returns:
            Texto do campo
        """
        return self._entry.get_text()


    def set_text(self, text: str):
        """
        Define o texto do campo de entrada.

        Args:
            text: Texto a ser definido
        """
        self._entry.set_text(text)


    def clear_entry(self):
        """Limpa o campo de entrada."""
        self.set_text('')


    def freeze(self, frozen: bool = True):
        """
        Congela/descongela o widget (desabilita interação).

        Args:
            frozen: True para congelar, False para descongelar
        """
        self._entry.set_sensitive(not frozen)

        if not frozen:
            self._update_button_state()
        else:
            self._button.set_sensitive(False)


    def focus_entry(self):
        """Coloca o foco no campo de entrada."""
        self._entry.grab_focus()


    def select_all(self):
        """Seleciona todo o texto do campo."""
        self._entry.select_region(0, -1)

    # ============== Métodos privados ==============

    def _update_button_state(self):
        """Atualiza o estado do botão baseado no texto."""
        has_text = len(self._entry.get_text()) > 0
        self._button.set_sensitive(has_text)


    def _is_frozen(self) -> bool:
        """
        Verifica se o widget está congelado.

        Returns:
            True se está congelado
        """
        return not self._entry.get_sensitive()

File: src/search_stock.rs
// search_stock.rs
//
// Copyright 2025 Flávio de Vasconcellos Corrêa
//
// SPDX-License-Identifier: GPL-3.0-or-later

use gtk4::prelude::*;
use gtk4::subclass::prelude::*;
use gtk4::{glib, CompositeTemplate};

mod imp {
    use super::*;
    use gtk4::TemplateChild;

    #[derive(Debug, Default, CompositeTemplate)]
    #[template(resource = "/com/github/sheepfarm/merkato/search_stock.ui")]
    pub struct SearchStock {
        #[template_child(id = "_entry")]
        pub entry: TemplateChild<gtk4::SearchEntry>,
        #[template_child(id = "_button")]
        pub button: TemplateChild<gtk4::Button>,
    }

    #[glib::object_subclass]
    impl ObjectSubclass for SearchStock {
        const NAME: &'static str = "MerkatoSearchStock";
        type Type = super::SearchStock;
        type ParentType = gtk4::Box;

        fn class_init(klass: &mut Self::Class) {
            klass.bind_template();
        }

        fn instance_init(obj: &glib::subclass::InitializingObject<Self>) {
            obj.init_template();
        }
    }

    impl ObjectImpl for SearchStock {
        fn constructed(&self) {
            self.parent_constructed();
            let entry = self.entry.clone();
            let button = self.button.clone();

            // Enable/disable button based on entry content
            entry.connect_changed(glib::clone!(
                #[weak]
                button,
                move |e| {
                    button.set_sensitive(!e.text().is_empty());
                }
            ));
            button.set_sensitive(false);
        }
    }

    impl WidgetImpl for SearchStock {}
    impl BoxImpl for SearchStock {}
}

glib::wrapper! {
    pub struct SearchStock(ObjectSubclass<imp::SearchStock>)
        @extends gtk4::Box, gtk4::Widget,
        @implements gtk4::Buildable, gtk4::ConstraintTarget, gtk4::Orientable;
}

impl SearchStock {
    pub fn new() -> Self {
        glib::Object::new()
    }

    pub fn text(&self) -> String {
        self.imp().entry.text().to_string()
    }

    pub fn set_text(&self, text: &str) {
        self.imp().entry.set_text(text);
    }

    pub fn set_frozen(&self, frozen: bool) {
        self.imp().entry.set_sensitive(!frozen);
        self.imp().button.set_sensitive(!frozen && !self.imp().entry.text().is_empty());
    }

    pub fn connect_search<F: Fn(String) + 'static>(&self, f: F) {
        let entry = self.imp().entry.clone();
        let button = self.imp().button.clone();

        let f = std::rc::Rc::new(f);
        let f2 = f.clone();

        button.connect_clicked(glib::clone!(
            #[weak]
            entry,
            move |_| {
                let text = entry.text().to_string();
                if !text.is_empty() {
                    f(text);
                }
            }
        ));

        self.imp().entry.connect_activate(move |e| {
            let text = e.text().to_string();
            if !text.is_empty() {
                f2(text);
            }
        });
    }
}

impl Default for SearchStock {
    fn default() -> Self {
        Self::new()
    }
}
